packages/protocol/lib/bytecode-foundry.ts

```diff
@@ -0,0 +1,213 @@
+/* eslint-disable max-classes-per-file: 0 */
+/*
+ * Foundry-specific bytecode utilities for handling library linking.
+ * 
+ * Foundry uses hashed library placeholders in the format __$<34-char-hex>$__
+ * instead of Truffle's name-based placeholders __LibraryName_______________.
+ * 
+ * The hash is computed as: keccak256("sourcePath:LibraryName").substring(2, 36)
+ */
+
+import { TextEncoder } from 'util'
+import { keccak256, toHex } from 'viem'
+
+// Reuse metadata stripping from the original bytecode.ts
+export { stripMetadata, verifyAndStripLibraryPrefix } from './bytecode'
+
+// Maps library names to their onchain addresses (formatted without "0x" prefix).
+export interface LibraryLinksFoundry {
+  [name: string]: string
+}
+
+// Maps library names to their source file paths (needed for hash computation)
+export interface LibrarySourcePaths {
+  [name: string]: string
+}
+
+/**
+ * Computes the Foundry library placeholder hash.
+ * Hash = keccak256("sourcePath:libraryName").substring(2, 36)
+ * 
+ * @param sourcePath - The source file path (e.g., "contracts/common/linkedlists/IntegerSortedLinkedList.sol")
+ * @param libraryName - The library name (e.g., "IntegerSortedLinkedList")
+ * @returns 34-character hex string (without 0x prefix)
+ */
+export const computeFoundryLibraryHash = (sourcePath: string, libraryName: string): string => {
+  const stringToHash = `${sourcePath}:${libraryName}`
+  const hashed = keccak256(toHex(new TextEncoder().encode(stringToHash)))
+  return hashed.substring(2, 2 + 34) // 34 chars
+}
+
+/**
+ * Links libraries in Foundry bytecode by replacing hashed placeholders with addresses.
+ * 
+ * @param bytecode - The bytecode containing __$<hash>$__ placeholders
+ * @param libraryLinks - Map of library names to their deployed addresses
+ * @param librarySourcePaths - Map of library names to their source file paths
+ * @returns Bytecode with placeholders replaced by addresses
+ */
+export const linkLibrariesFoundry = (
+  bytecode: string,
+  libraryLinks: LibraryLinksFoundry,
+  librarySourcePaths: LibrarySourcePaths
+): string => {
+  let linkedBytecode = bytecode
+
+  Object.keys(libraryLinks).forEach((libraryName) => {
+    const sourcePath = librarySourcePaths[libraryName]
+    if (!sourcePath) {
+      console.warn(`Warning: No source path found for library ${libraryName}, skipping linking`)
+      return
+    }
+
+    const placeholderHash = computeFoundryLibraryHash(sourcePath, libraryName)
+    const placeholder = `__\\$${placeholderHash}\\$__`
+
+    // Use g flag to replace all occurrences
+    const regex = new RegExp(placeholder, 'g')
+    linkedBytecode = linkedBytecode.replace(regex, libraryLinks[libraryName])
+  })
+
+  return linkedBytecode
+}
+
+const ADDRESS_LENGTH = 40
+
+/**
+ * Class for tracking positions of library placeholders in Foundry bytecode.
+ * Uses linkReferences from Foundry artifacts directly instead of regex parsing.
```

Comment:
```suggestion
 * Uses linkReferences from Foundry artifacts.
```
This is useful context in this PR, but not really a necessary comment in the future.